Ncnews 11

caolanhamilton opened the pull request:
NC News task 11

Added the ability to sort, order and filter articles as a query of /api/articles/ in models.js

Added additional tests to check:
- Returns article sorted by any valid column in either ascending or descending 
- Returns articles only with the queried topic 
- Returns an empty array when there are no articles for a valid topic 
- Return 400 bad request when passed an order value that is not asc or desc 
- Return 400 bad request when passed a sort query that is not a valid column 
- Return 400 bad request when passed a topic filter query that is not a valid topic

popatre commented on models/articles.model.js:
```diff
@@ -23,14 +23,31 @@ exports.addVoteToArticle = (articleID, votesToAmendBy) => {
+   
+    let queryString = `SELECT articles.author, articles.title, articles.article_id, articles.topic, articles.created_at, articles.votes, CAST(COUNT(comment_id) AS integer) AS comment_count FROM articles LEFT JOIN comments ON articles.article_id = comments.article_id `
+  
+    if(!['mitch', 'cats', 'paper', undefined].includes(topic)) {
```
This will cause you problems. 
1. The dev data has different topics. At the moment, you wont be able to find a topic in the dev data. 
2. it's not dynamics if you add a new topic. I think an extra ticket is the option to add topics. At the moment, they will be able to add a topic, but not use it. 

You'll need to query the db, to see what the current topics are/allowed.

popatre commented on __tests__/app.test.js:
```diff
@@ -194,6 +194,85 @@ describe('GET /api/articles', () => {
       })
     })
   });
+  test('200: Returns article sorted by any valid column in either ascending or descending', () => {
```
Is there a test for descending?

popatre commented on __tests__/app.test.js:
```diff
@@ -194,6 +194,85 @@ describe('GET /api/articles', () => {
+      expect(body.msg).toEqual('Invalid order query')
+    })
+  })
+  test('400: Return bad request when passed a sort query that is not a valid column', () => {
```
This would be 404 - valid but not found topic. 

another test - 400 - invalid topic - would probably get handled by psql.

popatre reviewed (changes requested):
Some tweaks and tests to make

caolanhamilton commented:
Fixed all things suggested:
-Made topics dynamic by making a request to fetchTopics and checking the topic exists instead of hard coding topics
-Added explicit test for descending order
-Added tests for 404 topic not found in topic table & 400 topic not valid topic type e.g number

bainesface commented on __tests__/app.test.js:
```diff
@@ -194,6 +194,115 @@ describe('GET /api/articles', () => {
+        })
+      })
+    })
+  });
```
Hey Caolan ... In order to keep your code as DRY as possible, you don't need to test the structure of all the articles for each individual query request. Just once will suffice and then you can test the behaviour of each query without. For example, using jest sorted to test that they are all sorted by the right column in the correct order

bainesface commented on models/articles.model.js:
```diff
@@ -23,14 +23,36 @@ exports.addVoteToArticle = (articleID, votesToAmendBy) => {
+        }
+        
+    }
+  
```
Be careful as you are still opening yourself up to sql injection here as you are building up a string to be passed directly as a query to the db. USe parameterised queries here with the $1 syntax

caolanhamilton commented:
Merging after approval and suggested refactors by Dav R on NC Help zoom
